### PR TITLE
Add binary chart

### PR DESCRIPTION
**Pull request by Terdious**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front)
- [x] Is the linter passing? (`npm run eslint` on both front)
- [x] Did you run prettier? (`npm run prettier` on both front)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to the community ([french forum](https://community.gladysassistant.com/)/[english forum](https://en-community.gladysassistant.com/)) for testing before merging.
- [ ] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [ ] Did you add fake requests data for the demo mode (`front/src/config/demo.js`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Taken from the original PR: #Binary graph #1948 
Thanks to @callemand  for all the work done

### CODE REVIEW

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
I created a device with 2 states in DB: opened at 3:32 in the morning, then closed at 7:32 in the morning : 

Screenshot 2024-06-07 at 09 38 48

Then, I open Gladys and create a binary chart for this device : 

Screenshot 2024-06-07 at 09 40 39

When I display the dashboard, the UI completely freezes and uses 100% CPU. I'm unable to do anything on the UI.

IMO there is something in the front that loops too much and blocks the JS

File: front/src/components/boxs/chart/ApexChartTimelineOptions.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      lines.forEach((line, index) => {
+        const tspan = document.createElementNS('http://www.w3.org/2000/svg', 'tspan');
+        if (line.length > 15) {
+          line = `${line.substring(0, 15)}...`;
+        }
+        tspan.setAttribute('x', text.getAttribute('x'));
+        tspan.setAttribute('dy', index === 0 ? '-0.4em' : '1.2em');
+        tspan.setAttribute('font-size', fontSize);
+        tspan.textContent = line;
+        tspan.title = line;
```
Comment (Terdious) [inline front/src/components/boxs/chart/ApexChartTimelineOptions.js]: Delete `tspan.title = line;`. Not used

File: front/src/components/boxs/chart/ApexChartComponent.jsx [hunk 2]
```diff
@@ -123,9 +184,31 @@ class ApexChartComponent extends Component {
...
+      height = 200;
+    }
+    const options = getApexChartTimelineOptions({
+      height,
+      series: this.props.series,
+      displayAxes: this.props.display_axes,
+      colors: mergeArray(this.props.colors, DEFAULT_COLORS),
+      locales: [fr, en, de],
+      defaultLocale: this.props.user.language
+    });
```
Comment (Terdious) [inline front/src/components/boxs/chart/ApexChartComponent.jsx:194]: Reorder options like other chart types

File: front/src/components/boxs/chart/Chart.jsx [hunk 3]
```diff
@@ -141,32 +141,87 @@ class Chartbox extends Component {
...
+                  serie0.data.push(newData);
+                } else {
+                  serie1.data.push(newData);
+                }
+                lastChangeTime = Math.round(new Date(value.created_at).getTime() / 1000) * 1000;
+                previousValue = value.value;
+              }
+            });
+
+            // Add the last value if it is different from the previous one
```
Comment (Terdious) [inline front/src/components/boxs/chart/Chart.jsx]: comment to delete

File: front/src/components/boxs/chart/EditChart.jsx [hunk 4]
```diff
@@ -298,28 +362,30 @@ class EditChart extends Component {
...
+                  </option>
+                </select>
+              </div>
+            )}
             <div class="form-group">
               <label>
                 <Text id="dashboard.boxes.chart.defaultInterval" />
               </label>
-              <select onChange={this.updateDefaultInterval} class="form-control" value={props.box.interval}>
+              <select onChange={this.updateDefaultInterval} className="form-control" value={props.box.interval}>
```
Comment (Terdious) [inline front/src/components/boxs/chart/EditChart.jsx]: Why change class to className?

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Hey @Terdious ! Thanks for the changes :) 

I have a few bugs when trying it: 

- Text is inside the "bar" : 

Screenshot 2024-06-13 at 10 00 52

- I get a freeze when saving the dashboard with a binary chart on a power plug

Here is the error : 

Screenshot 2024-06-13 at 10 03 12

The payload : 

Screenshot 2024-06-13 at 10 05 10


- If I select a binary device, then unselect it, then select a non-binary device, I get a weird state where the chart type is unknown : 

Before: 

Screenshot 2024-06-13 at 10 06 34

After : 

Screenshot 2024-06-13 at 10 06 49

- When I'm using Gladys in English, the text inside the popup is in french:

Screenshot 2024-06-13 at 10 08 45

- On mobile, the popup is outside of the screen : 

IMG_8681

Thanks for this PR 🙂

**Comment by Terdious** [comment]
> * Text is inside the "bar" :

I'm sorry but I tried to reproduce and despite zooming in / out changing resolution, etc. I never managed to get your result. On the other hand, before my previous modification I had the same problem, but I corrected it before your review

> * I get a freeze when saving the dashboard with a binary chart on a power plug

I added a `console.log(deviceFeature);` because I can't seem to reproduce. I tested all types of binary, tasmota and mqtt (including a power plug). And I have no problems with the units (everything is 'null'). Could you send me the logs?

> * If I select a binary device, then unselect it, then select a non-binary device, I get a weird state where the chart type is unknown 

Corrected

> * When I'm using Gladys in English, the text inside the popup is in french

Corrected


> * On mobile, the popup is outside of the screen

Corrected. Display top left box

**Comment by Pierre-Gilles** [comment]
@Terdious Thanks for the fixes! 

Here is the payload that causes the freeze: 

```
{"id":"e499dd5e-6482-4cbd-a73e-159d53c9b402","name":"Graph","selector":"graph","type":"main","visibility":"private","user_id":"275faa00-8a9c-4747-8fbe-417ddb966b16","created_at":"2024-06-10T08:20:20.784Z","updated_at":"2024-06-17T11:30:17.089Z","boxes":[[{"type":"chart","device_features":["mqtt-priseonoff"],"units":[null],"title":"Binaire","chart_type":"","interval":"last-day","display_axes":true}],[],[]]}
```

The response is 422 : 

```
{
    "status": 422,
    "code": "UNPROCESSABLE_ENTITY",
    "properties": [
        {
            "message": "\"[0][0].chart_type\" is not allowed to be empty",
            "attribute": "boxes",
            "value": [
                [
                    {
                        "type": "chart",
                        "device_features": [
                            "mqtt-priseonoff"
                        ],
                        "units": [
                            null
                        ],
                        "title": "Binaire",
                        "chart_type": "",
                        "interval": "last-day",
                        "display_axes": true
                    }
                ],
                [],
                []
            ],
            "type": "Validation error"
        }
    ]
}
```

It's just because I didn't select the "chart type" ^^ It's maybe not related to this PR, but I wonder if we could make a fix to avoid freezing in this situation... 

I have another feedback, now the date in english is not correct: 

Screenshot 2024-06-17 at 13 29 13

**Comment by Terdious** [comment]
> @TerdiousMerci pour les correctifs !
> 
> Voici la charge utile qui provoque le gel :
> 
> ```
> {"id":"e499dd5e-6482-4cbd-a73e-159d53c9b402","name":"Graph","selector":"graph","type":"main","visibility":"private","user_id":"275faa00-8a9c-4747-8fbe-417ddb966b16","created_at":"2024-06-10T08:20:20.784Z","updated_at":"2024-06-17T11:30:17.089Z","boxes":[[{"type":"chart","device_features":["mqtt-priseonoff"],"units":[null],"title":"Binaire","chart_type":"","interval":"last-day","display_axes":true}],[],[]]}
> ```
> 
> La réponse est 422 :
> 
> ```
> {
>     "status": 422,
>     "code": "UNPROCESSABLE_ENTITY",
>     "properties": [
>         {
>             "message": "\"[0][0].chart_type\" is not allowed to be empty",
>             "attribute": "boxes",
>             "value": [
>                 [
>                     {
>                         "type": "chart",
>                         "device_features": [
>                             "mqtt-priseonoff"
>                         ],
>                         "units": [
>                             null
>                         ],
>                         "title": "Binaire",
>                         "chart_type": "",
>                         "interval": "last-day",
>                         "display_axes": true
>                     }
>                 ],
>                 [],
>                 []
>             ],
>             "type": "Validation error"
>         }
>     ]
> }
> ```
> 
> C'est juste parce que je n'ai pas sélectionné le "type de graphique" ^^ Ce n'est peut-être pas lié à ce PR, mais je me demande si nous pourrions faire un correctif pour éviter de geler dans cette situation...
> 
> J'ai un autre retour, maintenant la date en anglais n'est pas correcte :
> 
> Capture d'écran 2024-06-17 à 13 h 29 min 13 s

Yes of course no worries, so you want us to put a lock to prohibit saving? Or to put an error text in place of the graph?

Edit: 
will this be suitable?
image

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for the changes! It's great ! 

I have 2 feedbacks:

I still have the date issue in english: 

Screenshot 2024-08-30 at 15 36 20

It seems the Websocket refresh is not working as with the other types of charts?

**Comment by Terdious** [comment]
> Thanks for the changes! It's great !
> 
> I have 2 feedbacks:
> 
> I still have the date issue in english:
> 
> Screenshot 2024-08-30 at 15 36 20

Indeed, I had not retested. It is good for the 3 languages:
imageimageimage

> 
> It seems the Websocket refresh is not working as with the other types of charts?

If you're talking about the automatic data update, I have no problem on my side, update without page refresh on my sensor every 30s...

**Comment by Pierre-Gilles** [comment]
Thanks for the changes! Indeed, for the auto refresh it's working on my side too. I think it's because the bar was so small I didn't see it, but in the request tab I can see the auto refresh working :)

File: front/src/components/boxs/chart/ApexChartComponent.jsx [hunk 5]
```diff
@@ -123,9 +174,32 @@ class ApexChartComponent extends Component {
...
   };
+  getTimelineChartOptions = () => {
+    let height;
+    if (this.props.size === 'small' && !this.props.display_axes) {
+      height = 40;
+    } else if (this.props.size === 'big' && !this.props.display_axes) {
+      height = 80;
+    } else {
+      // 95 is the height display of the timeline chart when there is no additional height
+      height = 95 + this.props.heightAdditional;
```
Comment (Pierre-Gilles) [inline front/src/components/boxs/chart/ApexChartComponent.jsx]: additionalHeight

File: front/src/components/boxs/chart/ApexChartTimelineOptions.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
+const addYAxisStyles = () => {
+  const yAxisLabel = document.querySelectorAll('.apexcharts-yaxis-label');
```
Comment (Pierre-Gilles) [inline front/src/components/boxs/chart/ApexChartTimelineOptions.js:2]: Where is this coming from?

ApexChart documentation?

File: server/test/lib/device/device.getDeviceFeaturesAggregates.test.js [hunk 7]
```diff
@@ -131,3 +167,86 @@ describe('Device.getDeviceFeaturesAggregates', function Describe() {
...
+    const deviceInstance = new Device(event, {}, stateManager, {}, {}, variable, job);
+    const { values, device, deviceFeature } = await deviceInstance.getDeviceFeaturesAggregates(
+      'binary-feature',
+      24 * 60,
+      300,
+    );
+    expect(values).to.have.lengthOf(300);
+    expect(device).to.have.property('name');
+    expect(deviceFeature).to.have.property('name');
+    // Vérifier que les valeurs sont bien des changements d'état
```
Comment (Pierre-Gilles) [inline server/test/lib/device/device.getDeviceFeaturesAggregates.test.js]: english please!

File: front/src/components/boxs/chart/Chart.jsx [hunk 8]
```diff
@@ -133,40 +137,82 @@ class Chartbox extends Component {
     }
     await this.setState({ loading: true });
     try {
+      const maxStates = 300;
```
Comment (Pierre-Gilles) [inline front/src/components/boxs/chart/Chart.jsx:140]: So in the end maxStates is 300 for everything?

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Hey! A few feedbacks on the code, nothing serious :)

Otherwise, tested locally and it works great! Great work 👏

File: front/src/components/boxs/chart/Chart.jsx (hunk 8 above)
Comment (Terdious) [reply front/src/components/boxs/chart/Chart.jsx:140]: Yes, 300 is the max in binary where everything goes well in display (no lag). As I had put it for the other graphs we could have put 3000, but as it is not the object of the PR and that it works very well for all, I preferred to put a single value for all the graphs for the moment, we will be able to go up in a separate PR with a greater control of tests on the other type of graph. Knowing that I am working on the expand ... to see?

File: front/src/components/boxs/chart/ApexChartTimelineOptions.js (hunk 6 above)
Comment (Terdious) [reply front/src/components/boxs/chart/ApexChartTimelineOptions.js:2]: No, I found this solution on stackoverflow and on apexcharts github

**Review by Pierre-Gilles — APPROVED** [review]
Good for me now! Thanks for all the good work :)

**Pull request merged**